File: docs/cvss31.js
/* CVSS v3.1 Basis-Score-Rechner – 1:1-Port von pipeline/vt_pipeline/cvss31.py
   (FIRST-Spezifikation Abschnitt 7, Roundup nach Appendix A).
   Muss dieselben Ergebnisse liefern wie computed_score in cves.json. */
(function (global) {
  'use strict';

  const METRICS = ['AV', 'AC', 'PR', 'UI', 'S', 'C', 'I', 'A'];
  const ALLOWED = {
    AV: ['N', 'A', 'L', 'P'], AC: ['L', 'H'], PR: ['N', 'L', 'H'], UI: ['N', 'R'],
    S: ['U', 'C'], C: ['N', 'L', 'H'], I: ['N', 'L', 'H'], A: ['N', 'L', 'H'],
  };
  const W_AV = { N: 0.85, A: 0.62, L: 0.55, P: 0.2 };
  const W_AC = { L: 0.77, H: 0.44 };
  const W_PR_U = { N: 0.85, L: 0.62, H: 0.27 };
  const W_PR_C = { N: 0.85, L: 0.68, H: 0.5 };
  const W_UI = { N: 0.85, R: 0.62 };
  const W_CIA = { H: 0.56, L: 0.22, N: 0.0 };

  function parseVector(vector) {
    const m = /^CVSS:3\.[01]\/(.+)$/.exec(String(vector).trim());
    if (!m) throw new Error('Kein CVSS-3.x-Vektor: ' + vector);
    const parts = {};
    for (const token of m[1].split('/')) {
      const i = token.indexOf(':');
      if (i < 0) throw new Error('Ungültiges Token ' + token);
      const key = token.slice(0, i), val = token.slice(i + 1);
      if (ALLOWED[key]) {
        if (!ALLOWED[key].includes(val)) throw new Error('Ungültiger Wert ' + key + ':' + val);
        parts[key] = val;
      }
    }
    const missing = METRICS.filter(k => !(k in parts));
    if (missing.length) throw new Error('Fehlende Metriken ' + missing.join(','));
    const out = {};
    for (const k of METRICS) out[k] = parts[k];
    return out;
  }

  function formatVector(metrics) {
    return 'CVSS:3.1/' + METRICS.map(k => k + ':' + metrics[k]).join('/');
  }

  // Roundup nach Appendix A – vermeidet Float-Artefakte.
  function roundup(x) {
    const intInput = Math.round(x * 100000);
    if (intInput % 10000 === 0) return intInput / 100000.0;
    return (Math.floor(intInput / 10000) + 1) / 10.0;
  }

  function baseScore(m) {
    const sChanged = m.S === 'C';
    const iss = 1 - (1 - W_CIA[m.C]) * (1 - W_CIA[m.I]) * (1 - W_CIA[m.A]);
    const impact = sChanged
      ? 7.52 * (iss - 0.029) - 3.25 * Math.pow(iss - 0.02, 15)
      : 6.42 * iss;
    const prW = (sChanged ? W_PR_C : W_PR_U)[m.PR];
    const exploitability = 8.22 * W_AV[m.AV] * W_AC[m.AC] * prW * W_UI[m.UI];
    if (impact <= 0) return 0.0;
    if (sChanged) return roundup(Math.min(1.08 * (impact + exploitability), 10));
    return roundup(Math.min(impact + exploitability, 10));
  }

  function severity(score) {
    if (score === 0) return 'NONE';
    if (score < 4.0) return 'LOW';
    if (score < 7.0) return 'MEDIUM';
    if (score < 9.0) return 'HIGH';
    return 'CRITICAL';
  }

  function diffVectors(a, b) {
    const pa = parseVector(a), pb = parseVector(b);
    return METRICS.filter(k => pa[k] !== pb[k]);
  }

  global.CVSS31 = { METRICS, ALLOWED, parseVector, formatVector, roundup, baseScore, severity, diffVectors };
})(window);

File: pipeline/vt_pipeline/cvss31.py
"""CVSS v3.1 Basis-Score-Rechner (Spezifikation FIRST, Abschnitt 7).

Rein, ohne Abhängigkeiten. Wird in der Pipeline benutzt, um jeden Vektor
gegen den offiziellen Score zu prüfen. Das TypeScript-Gegenstück im Web
muss dieselben Testvektoren (tests/test_cvss31.py) bestehen.
"""
from __future__ import annotations

import math
import re

METRICS = ("AV", "AC", "PR", "UI", "S", "C", "I", "A")

ALLOWED = {
    "AV": ("N", "A", "L", "P"),
    "AC": ("L", "H"),
    "PR": ("N", "L", "H"),
    "UI": ("N", "R"),
    "S": ("U", "C"),
    "C": ("N", "L", "H"),
    "I": ("N", "L", "H"),
    "A": ("N", "L", "H"),
}

_W_AV = {"N": 0.85, "A": 0.62, "L": 0.55, "P": 0.2}
_W_AC = {"L": 0.77, "H": 0.44}
_W_PR_U = {"N": 0.85, "L": 0.62, "H": 0.27}
_W_PR_C = {"N": 0.85, "L": 0.68, "H": 0.5}
_W_UI = {"N": 0.85, "R": 0.62}
_W_CIA = {"H": 0.56, "L": 0.22, "N": 0.0}

_VECTOR_RE = re.compile(r"^CVSS:3\.[01]/(.+)$")


def parse_vector(vector: str) -> dict[str, str]:
    """Zerlegt 'CVSS:3.1/AV:N/AC:L/...' in ein Dict der acht Basismetriken.

    Temporal-/Environmental-Metriken werden ignoriert. Wirft ValueError bei
    fehlenden oder ungültigen Basismetriken.
    """
    m = _VECTOR_RE.match(vector.strip())
    if not m:
        raise ValueError(f"Kein CVSS-3.x-Vektor: {vector!r}")
    parts: dict[str, str] = {}
    for token in m.group(1).split("/"):
        if ":" not in token:
            raise ValueError(f"Ungültiges Token {token!r} in {vector!r}")
        key, val = token.split(":", 1)
        if key in ALLOWED:
            if val not in ALLOWED[key]:
                raise ValueError(f"Ungültiger Wert {key}:{val} in {vector!r}")
            parts[key] = val
    missing = [k for k in METRICS if k not in parts]
    if missing:
        raise ValueError(f"Fehlende Metriken {missing} in {vector!r}")
    return {k: parts[k] for k in METRICS}


def format_vector(metrics: dict[str, str]) -> str:
    return "CVSS:3.1/" + "/".join(f"{k}:{metrics[k]}" for k in METRICS)


def roundup(x: float) -> float:
    """Roundup nach CVSS-3.1-Spezifikation (Appendix A), vermeidet Float-Artefakte."""
    int_input = round(x * 100000)
    if int_input % 10000 == 0:
        return int_input / 100000.0
    return (math.floor(int_input / 10000) + 1) / 10.0


def base_score(metrics: dict[str, str]) -> float:
    s_changed = metrics["S"] == "C"
    iss = 1 - (1 - _W_CIA[metrics["C"]]) * (1 - _W_CIA[metrics["I"]]) * (1 - _W_CIA[metrics["A"]])
    if s_changed:
        impact = 7.52 * (iss - 0.029) - 3.25 * (iss - 0.02) ** 15
    else:
        impact = 6.42 * iss
    pr_w = (_W_PR_C if s_changed else _W_PR_U)[metrics["PR"]]
    exploitability = 8.22 * _W_AV[metrics["AV"]] * _W_AC[metrics["AC"]] * pr_w * _W_UI[metrics["UI"]]
    if impact <= 0:
        return 0.0
    if s_changed:
        return roundup(min(1.08 * (impact + exploitability), 10))
    return roundup(min(impact + exploitability, 10))


def score_from_vector(vector: str) -> float:
    return base_score(parse_vector(vector))


def severity(score: float) -> str:
    if score == 0:
        return "NONE"
    if score < 4.0:
        return "LOW"
    if score < 7.0:
        return "MEDIUM"
    if score < 9.0:
        return "HIGH"
    return "CRITICAL"


def diff_vectors(a: str, b: str) -> list[str]:
    """Liste der Metriken, in denen sich zwei Vektoren unterscheiden."""
    pa, pb = parse_vector(a), parse_vector(b)
    return [k for k in METRICS if pa[k] != pb[k]]
